lxd/storage/drivers/driver_cephfs.go

```diff
@@ -0,0 +1,977 @@
+	if volType != VolumeTypeCustom {
+		return fmt.Errorf("Volume type not supported")
+	}
+
+	sourcePath := GetVolumeMountPath(d.name, volType, volName)
+	cephSnapPath := filepath.Join(sourcePath, ".snap", snapshotName)
+
+	err := os.RemoveAll(cephSnapPath)
+	if err != nil {
+		return err
+	}
+
+	// Make the snapshot path a symlink
+	snapPath := GetVolumeMountPath(d.name, volType, GetSnapshotVolumeName(volName, snapshotName))
+
+	err = os.RemoveAll(snapPath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *cephfs) RenameVolumeSnapshot(volType VolumeType, volName string, snapshotName string, newSnapshotName string, op *operations.Operation) error {
+	if volType != VolumeTypeCustom {
+		return fmt.Errorf("Volume type not supported")
+	}
+
+	sourcePath := GetVolumeMountPath(d.name, volType, volName)
+	oldCephSnapPath := filepath.Join(sourcePath, ".snap", snapshotName)
+	newCephSnapPath := filepath.Join(sourcePath, ".snap", newSnapshotName)
+
+	err := os.Rename(oldCephSnapPath, newCephSnapPath)
+	if err != nil {
+		return err
+	}
+
+	// Re-generate the snapshot symlink
+	oldPath := GetVolumeMountPath(d.name, volType, GetSnapshotVolumeName(volName, snapshotName))
+	err = os.Remove(oldPath)
+	if err != nil {
+		return err
+	}
+
+	newPath := GetVolumeMountPath(d.name, volType, GetSnapshotVolumeName(volName, newSnapshotName))
+	err = os.Symlink(newCephSnapPath, newPath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *cephfs) VolumeSnapshots(volType VolumeType, volName string, op *operations.Operation) ([]string, error) {
+	if volType != VolumeTypeCustom {
+		return nil, fmt.Errorf("Volume type not supported")
+	}
+
+	snapshotDir, err := GetVolumeSnapshotDir(d.name, volType, volName)
+	if err != nil {
+		return nil, err
+	}
+
+	snapshots := []string{}
+
+	ents, err := ioutil.ReadDir(snapshotDir)
+	if err != nil {
+		// If the snapshots directory doesn't exist, there are no snapshots.
+		if os.IsNotExist(err) {
+			return snapshots, nil
+		}
+
+		return nil, err
+	}
+
+	for _, ent := range ents {
+		fileInfo, err := os.Stat(filepath.Join(snapshotDir, ent.Name()))
+		if err != nil {
+			return nil, err
+		}
+
+		if !fileInfo.IsDir() {
+			continue
+		}
+
+		snapshots = append(snapshots, ent.Name())
+	}
+
+	return snapshots, nil
+}
+
+func (d *cephfs) RestoreVolume(vol Volume, snapshotName string, op *operations.Operation) error {
+	sourcePath := GetVolumeMountPath(d.name, vol.volType, vol.name)
+	cephSnapPath := filepath.Join(sourcePath, ".snap", snapshotName)
+
+	// Restore using rsync.
+	bwlimit := d.config["rsync.bwlimit"]
+	output, err := rsync.LocalCopy(cephSnapPath, vol.MountPath(), bwlimit, false)
+	if err != nil {
+		return fmt.Errorf("Failed to rsync volume: %s: %s", string(output), err)
+	}
+
+	return nil
+}
+
+func (d *cephfs) MigrateVolume(vol Volume, conn io.ReadWriteCloser, volSrcArgs migration.VolumeSourceArgs, op *operations.Operation) error {
```

Comment:
Same comment as CreateFromCopy, needs to change to match the fix I did on `dir`.